### PR TITLE
vehicles: use fake parts to fill gaps in turning vehicles

### LINKED ISSUE #5684: Angled vehicles develop holes [$15]
While all this vehicle renaissance is going on, does it seem feasable to tackle that problem where cars at a 45-degree angle can be entered diagonally between otherwise solid parts? I think you know what I mean, how if two boards are adjacent vertically, at an angle you'll cross between them somehow. Ideally it would check the vehicle structure and make the space beyond functionally impassable/opaque from only that side. It'd probably be a tough fix, but it ought to happen at some point.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1548800-angled-vehicles-develop-holes?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by robob27**
#### Summary
Features "vehicles: use fake parts to fill gaps in turning vehicles"

#### Purpose of change
(comments by original author)
Fixes #5684

As vehicles turn, they skew and open gaps in previously aligned lines of obstacles. This creates a weird situation where a vehicle that is driving in a cardinal direction fully protects its occupants, but a vehicle driving at any other angle can get hitchhikers.

#### Describe the solution
(comments by original author)
This is a cleaned up and simplified version of #41277 by mft-dev and he should get co-contributor credit.

Create a bunch of functions to return vpart_ranges with and without fake parts and active and inactive fake parts.

During vehicle::refresh(), find all the OBSTACLE parts on the left/right edges of the vehicle, and add another part 1 map tile to the left or right of that part. The new part is a fake part and is not counted for vehicle weight, the vehicle map cache, etc. Fake parts have an index value of the real part that they are mimicking (their parent), and the parent part has an index value of the fake part.

When the vehicle turns, a function finds all the fake parts that are no longer in line with their parents and changes them from inactive fakes to active fakes (and changes unneeded active fakes in inactive fakes). Active fakes are displayed and added to the map cache, so they block like of sight and movement. In all cases, damage to a fake part is applied to the fake part instead.

#### Describe alternatives you've considered
(comments by original author)
bool is_fake is probably redundant to having a parent_index > 0 but it's easy to reason with.

Various other solutions to the angle gap problem have been proposed, but this solution is the only solution that can be confined mostly the vehicle code instead of polluting every part of the game.

#### Testing
(comments by original author)
Several new unit tests were added. One drives the vehicle in a circle and checks that the number of active fake parts changes correctly. Another makes sure that a collision on an active fake part does not damage the fake part but does damage the parent part.

#### Additional context
[Short gif of a vehicle turning through 360 degrees](https://www.prismnet.com/~mlangsdo/CDDA-AngledGapsFinal.gif)

Most of these changes and this PR description were originally submitted in #42462 by @mlangsdorf and @mft-dev. This PR attempts to fix the bugs described by testers on the first PR in an attempt to push this one past the finish line. The OG authors deserve all of the credit for this work, I've done very little here.

I am still doing some testing. I will update this comment with the results of my tests. As of right now, the bugs described in the original PR seem to be resolved and I haven't noticed any new issues but there are tons of changes here and... who knows.

TODO:
- [ ] Autodrive testing especially with narrow spaces
- [ ] More testing with folding/unfolding

### CODE REVIEW

File: data/json/vehicles/test.json.rej [hunk 1]
```diff
@@ -0,0 +1,96 @@
+--- data/json/vehicles/test.json
```
Comment (BrettDong) [inline data/json/vehicles/test.json.rej:1]: Is this file added by accident?

Comment (robob27) [reply data/json/vehicles/test.json.rej:1]: Possibly, this is one of the files in the original commits by the looks of it. I'll get it out of there if it's not needed for anything.

**Comment by GoLoT** [comment]
The autodrive feature was (not so) recently improved by @eltank and while I'm not sure how it works exactly it's possible that the new parts being created while turning will mess up some of the calculations. Any chance you can do a few tests while autodriving through narrow spaces an/or around obstacles to make sure it still works?

**Comment by Tommy-os** [comment]
How does this work for 1 tile wide vehicles like bikes?

2 tile wide vehicles?

Why only left/right sides?

It doesn't seem like a good solution, it makes the vehicles much wider than they actually are. Diagonal movement should be forbidden through straight vehicle walls.

**Comment by GoLoT** [comment]
> It doesn't seem like a good solution, it makes the vehicles much wider than they actually are. Diagonal movement should be forbidden through straight vehicle walls.

While this solution is not perfect it's the only accepted solution, so it's either this or having monsters phase through walls when you turn a vehicle. Blocking diagonal visibility/movement through vehicle tiles has been proposed and while it can be done, it comes with its own issues. Mainly that it requires updating visibility, movement, pathfinding and lighting code with ad-hoc/hardcoded checks, making those classes aware of implementation details of the vehicle class and causing inconsistencies.

So unless movement/visibility between diagonal blockers (walls, vehicles or w/e) is restricted globally it won't be implemented. And this would be a really big change at many levels that I don't think will happen.

**Comment by Zireael07** [comment]
>  And this would be a really big change at many levels that I don't think will happen.

This is precisely why this is the only acceptable solution. The alternative is so HUGE that it was attempted once and quickly abandoned in favor of this.

**Comment by Tommy-os** [comment]
It's solving a rare edge case while introducing a myriad of new edge cases.

**Comment by GoLoT** [comment]
> It's solving a rare edge case while introducing a myriad of new edge cases.

FYI: Limiting diagonal movement/visibility only for vehicles would be way more impactful and add more corner cases than this.

**Comment by Inglonias** [comment]
Yeah. While this fixes an odd, though common situation, I am also nervous about the edge cases. Autodrive, though improved, is still not perfect. It doesn't crash into stuff anymore which is great, but it is kind of scared of the dark.

It's definitely worth bringing this PR up again, but mergers should be very cautious about this one.

**Comment by robob27** [comment]
Converting to draft until I have a chance to do the autodrive testing and fix the linting.

File: src/mapgen.cpp [hunk 2]
```diff
@@ -6766,6 +6766,7 @@ std::unique_ptr<vehicle> map::add_vehicle_to_map(
     if( needs_smashing ) {
         veh_to_add->smash( *this );
     }
+    veh->refresh();
```
Comment (kevingranade) [inline src/mapgen.cpp:6769]:
```suggestion
    veh_to_add->refresh();
```

File: src/vehicle.cpp [hunk 3]
```diff
@@ -3323,23 +3370,25 @@ itype_id vehicle::engine_fuel_current( int e ) const
...
-    for( const vehicle_part &vp : parts ) {
-        if( vp.is_tank() && vp.ammo_current() == ftype &&
-            vp.base.only_item().made_of( phase_id::LIQUID ) ) {
-            total_energy += vp.base.only_item().get_item_thermal_energy();
-            total_mass += to_gram( vp.base.only_item().weight() );
+    float total_mass = 0.0f; // g
+    for( const vpart_reference &vpr : get_all_parts() ) {
+        if( vpr.part().is_tank() && vpr.part().ammo_current() == ftype &&
+            vpr.part().base.only_item().made_of( phase_id::LIQUID ) ) {
+            total_energy += vpr.part().base.only_item().get_item_termal_energy();
```
Comment (kevingranade) [inline src/vehicle.cpp:3390]:
```suggestion
            total_energy += vpr.part().base.only_item().get_item_thermal_energy();
```

File: src/vehicle.cpp [hunk 4]
```diff
@@ -2816,9 +2850,10 @@ vehicle_part_with_feature_range<vpart_bitflags> vehicle::get_enabled_parts(
 std::vector<int> vehicle::all_parts_at_location( const std::string &location ) const
 {
     std::vector<int> parts_found;
-    for( size_t part_index = 0; part_index < parts.size(); ++part_index ) {
-        if( part_info( part_index ).location == location && !parts[part_index].removed ) {
-            parts_found.push_back( part_index );
+    auto all_parts = get_all_parts();
```
Comment (kevingranade) [inline src/vehicle.cpp:2853]:
```suggestion
    vehicle_part_range all_parts = get_all_parts();
```

**Comment by kevingranade** [comment]
Less than stellar outcome from finally getting around to testing autodrive, it doesn't work at all, as in when I triggered it it just halts immediately on open ground.
"Can't see a path forward."
"Auto-drive cancelled."

**Comment by kevingranade** [comment]
Ok it looks like something was weird about one particular save, when I rolled back to HEAD of master there was a similar problem.

**Comment by kevingranade** [comment]
With a fresh save, autodrive is working just fine, I might have had a weird corrupted vehicle.
After driving around a ton, I'm getting periodic spam of:
src/vehicle.cpp:1825 [bool vehicle::remove_part(int, RemovePartHandler&)] Tried to remove part 299 but only 156 parts!
Which I *think* is a symptom of trying to clean up after forming wreckage, but I'm not sure.

Also wreckage forming was failing (I think) because of vehicle::get_points() enumerating points including fake parts, then vehicle::parts_at() for one of those points returning an empty vector, that's easy to skip over though.

**Pull request closed without merging**